src/rdf4cpp/rdf/Dataset.hpp

```diff
@@ -0,0 +1,66 @@
+#ifndef RDF4CPP_DATASET_HPP
+#define RDF4CPP_DATASET_HPP
+
+#include <rdf4cpp/rdf/Graph.hpp>
+#include <rdf4cpp/rdf/Quad.hpp>
+#include <rdf4cpp/rdf/storage/tuple/DefaultDatasetBackend.hpp>
+#include <rdf4cpp/rdf/storage/tuple/IDatasetBackend.hpp>
+
+#include <memory>
+#include <utility>
+
+
+namespace rdf4cpp::rdf {
+
+class Graph;
+
+class Dataset {
+    using IDatasetBackend = ::rdf4cpp::rdf::storage::tuple::IDatasetBackend;
+    using NodeStorage = storage::node::NodeStorage;
+    friend class Graph;
+
+    std::shared_ptr<IDatasetBackend> dataset_backend_;
+
+    explicit Dataset(std::shared_ptr<IDatasetBackend> datasetBackend);
+
+public:
+    // TODO: allow to change default backend impl.
+    explicit Dataset(NodeStorage node_storage = NodeStorage::primary_instance());
+
+
+    void add(const Quad &quad);
+
+    [[nodiscard]] bool contains(const Quad &quad) const;
+
+    [[nodiscard]] query::SolutionSequence match(const query::QuadPattern &quad_pattern) const;
+
+    [[nodiscard]] size_t size() const;
+
+    [[nodiscard]] size_t size(const IRI &graph_name) const;
+
+    Graph graph(const IRI &graph_name);
+
+    Graph graph();
+
+    void add_ttl_file(const std::string &path);
```

Comment:
good for now, 
Would change that in future to read_from_stream()

Reply:
Yes, I agree. I'll add a todo to remove it later.